### PR TITLE
upgrade the repo to ts 4.7 part 1

### LINKED ISSUE #3550: [eslint-config] Need update for dependencies of eslint-config
## Summary
In my project we're using TypeScript 4.7.4 and also using the `@rushstack/eslint-config` package, but when I run linting, I get this warning:

```
=============

WARNING: You are currently running a version of TypeScript which is not officially supported by @typescript-eslint/typescript-estree.

You may find that it works just fine, or you may not.

SUPPORTED TYPESCRIPT VERSIONS: >=3.3.1 <4.7.0

YOUR TYPESCRIPT VERSION: 4.7.4

Please only submit bug reports when using the officially supported version.

=============
```

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@rushstack/eslint-config` version? | 2.6.2 |
| Operating system? | Windows |
| Would you consider contributing a PR? | No |
| TypeScript compiler version? | 4.7.4 |
| Node.js version (`node -v`)? | 16.9.1 |

### PR DESCRIPTION
**Pull request by BPapp-MS**
## Summary

Upgrades the packages of the repo from using TypeScript 4.6 to 4.7. Some packages that consume rigs from the feed will need to pull them down in a separate PR.
Fixes #3550 
## Details

Upgraded the explicit dependencies on typescript from ~4.6.3 to ~4.7.4 , and updated the common-versions.json accordingly. Addressed an error in rush-lib that started as a result of a new typescript constraint on checking the existence of functions. 

In eslint-config, the bump breaks compatibility with projects using versions of TypeScript under 4.7 , and for that reason I upgraded the peerDependency on TypeScript accordingly. Due to cyclic dependencies, some packages rely on older versions of the rig and therefore TypeScript, and thus eslint is disabled in these packages by renaming .eslintrc.js to ._eslintrc.js , which I'll undo when following up on this pr by consuming the newly published rigs. I considered instead relying on an older feed version of the eslint-config in these projects instead of the local workspace:* version, but @iclanton advised this would be unnecessary and less simple.

There was an undocumented behavior in main that the default export of index files would be scoped in API Extractor markdown  files as "@public" regardless of how they were scoped in the source file. This PR remedies that behavior, and to keep some default exported heft plugins public, again under the advisement of Ian, I changed "@internal" to "@public" in the corresponding index files, avoiding changes in the api.md files.

## How it was tested

Validated via full typescript build locally.

### CODE REVIEW

File: build-tests/api-extractor-scenarios/etc/defaultExportOfEntryPoint4/rollup.d.ts [hunk 1]
```diff
@@ -1,3 +1,4 @@
+/** @public */
```
Comment (iclanton) [inline build-tests/api-extractor-scenarios/etc/defaultExportOfEntryPoint4/rollup.d.ts:1]: Why'd this change?

File: build-tests/eslint-7-test/package.json [hunk 2]
```diff
@@ -13,8 +13,8 @@
     "@rushstack/eslint-config": "workspace:*",
     "@rushstack/heft": "workspace:*",
     "@types/node": "12.20.24",
-    "@typescript-eslint/parser": "~5.20.0",
+    "@typescript-eslint/parser": "5.30.3",
```
Comment (iclanton) [inline build-tests/eslint-7-test/package.json:16]: Dropping the `~`?

File: common/config/rush/common-versions.json [hunk 3]
```diff
@@ -76,7 +76,7 @@
 
       // The newest supported compiler, used by most build tests and used as the bundled compiler
       // engine for API Extractor.
-      "~4.5.2"
+      "~4.6.3"
```
Comment (iclanton) [inline common/config/rush/common-versions.json:79]: Line 72 should get updated.

Does this line even need to be here?

File: common/reviews/api/heft-dev-cert-plugin.api.md [hunk 4]
```diff
@@ -6,8 +6,10 @@
 
 import type { IHeftPlugin } from '@rushstack/heft';
 
-// @public (undocumented)
+// @internal (undocumented)
```
Comment (iclanton) [inline common/reviews/api/heft-dev-cert-plugin.api.md:9]: These should be public.

File: libraries/rush-lib/src/logic/setup/KeyboardLoop.ts [hunk 5]
```diff
@@ -12,6 +12,8 @@ import { AlreadyReportedError, InternalError } from '@rushstack/node-core-librar
 const ANSI_ESCAPE_SHOW_CURSOR: string = '\u001B[?25l';
 const ANSI_ESCAPE_HIDE_CURSOR: string = '\u001B[?25h';
 
+type OptionalProperty<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>> & Partial<Pick<T, K>>;
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/setup/KeyboardLoop.ts:15]: Unrelated change?

**Comment by iclanton** [comment]
This should get changed: https://github.com/microsoft/rushstack/blob/main/apps/heft/src/plugins/TypeScriptPlugin/TypeScriptBuilder.ts#L118

File: libraries/rush-lib/src/logic/setup/KeyboardLoop.ts (hunk 5 above)
Comment (BPapp-MS) [reply libraries/rush-lib/src/logic/setup/KeyboardLoop.ts:15]: This is to resolve the one typescript error that surfaced from the version bump- that there is a check later in the file for the existence of a required function of stdin's type, so I cast is as a type where that property is optional for that check

File: common/reviews/api/heft-dev-cert-plugin.api.md (hunk 4 above)
Comment (BPapp-MS) [reply common/reviews/api/heft-dev-cert-plugin.api.md:9]: okay I'll change the comments in the index files for these packages

File: build-tests/eslint-7-test/package.json (hunk 2 above)
Comment (BPapp-MS) [reply build-tests/eslint-7-test/package.json:16]: I'll fix that

File: build-tests/install-test-workspace/workspace/common/pnpm-lock.yaml [hunk 6]
```diff
@@ -1677,12 +1675,12 @@ packages:
     dev: true
 
   /validator/13.7.0:
-    resolution: {integrity: sha512-nYXQLCBkpJ8X6ltALua9dRrZDHVYxjJ1wgskNt1lH9fzGjs3tgojGSCBjmEPwkWS1y29+DrizMTW19Pr9uB2nw==}
+    resolution: {integrity: sha1-T5ZYuhO6jz2C7ogdNRZInqhcCFc=}
```
Comment (BPapp-MS) [inline build-tests/install-test-workspace/workspace/common/pnpm-lock.yaml:1678]: Is this change in hash expected?

File: build-tests/api-extractor-scenarios/etc/defaultExportOfEntryPoint4/rollup.d.ts (hunk 1 above)
Comment (BPapp-MS) [reply build-tests/api-extractor-scenarios/etc/defaultExportOfEntryPoint4/rollup.d.ts:1]: I think this is just a compiler difference

File: common/changes/@microsoft/api-extractor-model/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
```
Comment (iclanton) [inline common/changes/@microsoft/api-extractor-model/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:4]:
```suggestion
      "comment": "Upgrade TypeScript dependency to 4.7",
```

File: common/changes/@microsoft/api-extractor/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 8]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
```
Comment (iclanton) [inline common/changes/@microsoft/api-extractor/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:4]:
```suggestion
      "comment": "Upgrade TypeScript dependency to 4.7",
```

File: common/changes/@microsoft/rush/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 9]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
```
Comment (iclanton) [inline common/changes/@microsoft/rush/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:4]:
```suggestion
      "comment": "",
```

Comment (iclanton) [inline common/changes/@microsoft/rush/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:4]:
```suggestion
      "comment": "",
```

Consumers of this package won't care.

File: common/changes/@rushstack/eslint-plugin-packlets/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 10]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
+      "type": "minor",
```
Comment (iclanton) [inline common/changes/@rushstack/eslint-plugin-packlets/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:5]:
```suggestion
      "comment": "Upgrade TypeScript dependency to 4.7",
      "type": "minor",
```

File: common/changes/@rushstack/eslint-plugin-security/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 11]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
```
Comment (iclanton) [inline common/changes/@rushstack/eslint-plugin-security/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:4]:
```suggestion
      "comment": "Upgrade TypeScript dependency to 4.7",
```

File: common/changes/@rushstack/eslint-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 12]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
```
Comment (iclanton) [inline common/changes/@rushstack/eslint-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:4]:
```suggestion
      "comment": "Upgrade TypeScript dependency to 4.7",
```

File: common/changes/@rushstack/heft-config-file/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 13]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
+      "type": "minor",
```
Comment (iclanton) [inline common/changes/@rushstack/heft-config-file/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:5]:
```suggestion
      "comment": "",
      "type": "none",
```

This doesn't matter to consumers.

File: common/changes/@rushstack/heft-dev-cert-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 14]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
+      "type": "minor",
```
Comment (iclanton) [inline common/changes/@rushstack/heft-dev-cert-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:5]:
```suggestion
      "comment": "",
      "type": "none",
```

This doesn't matter to consumers.

File: common/changes/@rushstack/heft-jest-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 15]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
+      "type": "minor",
```
Comment (iclanton) [inline common/changes/@rushstack/heft-jest-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:5]:
```suggestion
      "comment": "",
      "type": "none",
```

This doesn't matter to consumers.

File: common/changes/@rushstack/heft-node-rig/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 16]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
```
Comment (iclanton) [inline common/changes/@rushstack/heft-node-rig/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:4]:
```suggestion
      "comment": "Upgrade TypeScript to 4.7",
```

File: common/changes/@rushstack/heft-sass-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 17]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
```
Comment (iclanton) [inline common/changes/@rushstack/heft-sass-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:4]:
```suggestion
      "comment": "Upgrade TypeScript to 4.7",
```

File: common/changes/@rushstack/heft-storybook-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 18]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
+      "type": "minor",
```
Comment (iclanton) [inline common/changes/@rushstack/heft-storybook-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:5]:
```suggestion
      "comment": "",
      "type": "none",
```

This doesn't matter to consumers.

File: common/changes/@rushstack/heft-web-rig/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 19]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
```
Comment (iclanton) [inline common/changes/@rushstack/heft-web-rig/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:4]:
```suggestion
      "comment": "Upgrade TypeScript to 4.7",
```

File: common/changes/@rushstack/heft-sass-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 20]
```diff
@@ -0,0 +1,11 @@
...
+      "type": "minor",
```
Comment (iclanton) [inline common/changes/@rushstack/heft-sass-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:5]:
```suggestion
      "comment": "",
      "type": "none",
```

This doesn't matter to consumers.

File: common/changes/@rushstack/heft-webpack4-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 21]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
+      "type": "minor",
```
Comment (iclanton) [inline common/changes/@rushstack/heft-webpack4-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:5]:
```suggestion
      "comment": "",
      "type": "none",
```

This doesn't matter to consumers.

File: common/changes/@rushstack/heft-webpack5-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 22]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
+      "type": "minor",
```
Comment (iclanton) [inline common/changes/@rushstack/heft-webpack5-plugin/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:5]:
```suggestion
      "comment": "",
      "type": "none",
```

This doesn't matter to consumers.

File: common/changes/@rushstack/heft/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 23]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
```
Comment (iclanton) [inline common/changes/@rushstack/heft/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:4]:
```suggestion
      "comment": "Update the highest supported version of TypeScript to 4.7",
```

File: common/changes/@rushstack/node-core-library/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 24]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
+      "type": "minor",
```
Comment (iclanton) [inline common/changes/@rushstack/node-core-library/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:5]:
```suggestion
      "comment": "",
      "type": "none",
```

This doesn't matter to consumers.

File: common/changes/@rushstack/rig-package/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 25]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
+      "type": "minor",
```
Comment (iclanton) [inline common/changes/@rushstack/rig-package/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:5]:
```suggestion
      "comment": "",
      "type": "none",
```

This doesn't matter to consumers.

File: common/changes/@rushstack/tree-pattern/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 26]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
+      "type": "minor",
```
Comment (iclanton) [inline common/changes/@rushstack/tree-pattern/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:5]:
```suggestion
      "comment": "",
      "type": "none",
```

This doesn't matter to consumers.

File: common/changes/@rushstack/ts-command-line/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 27]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7",
+      "type": "minor",
```
Comment (iclanton) [inline common/changes/@rushstack/ts-command-line/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:5]:
```suggestion
      "comment": "",
      "type": "none",
```

This doesn't matter to consumers.

File: common/config/rush/common-versions.json [hunk 28]
```diff
@@ -63,31 +61,21 @@
...
     "eslint": ["~7.30.0"],
-
     /**
      * For example, allow some projects to use an older TypeScript compiler
      * (in addition to whatever "usual" version is being used by other projects in the repo):
      */
     "typescript": [
-      // "~4.5.2" is the (inferred, not alternative) range used by most projects in this repo
-
       // The oldest supported compiler, used by build-tests/api-extractor-lib1-test
```
Comment (iclanton) [inline common/config/rush/common-versions.json:69]:
```suggestion
      // "~4.7.4" is the (inferred, not alternative) range used by most projects in this repo

      // The oldest supported compiler, used by build-tests/api-extractor-lib1-test
```

File: eslint/eslint-config/package.json [hunk 29]
```diff
@@ -24,23 +24,23 @@
   ],
   "peerDependencies": {
     "eslint": "^6.0.0 || ^7.0.0 || ^8.0.0",
-    "typescript": ">=3.0.0"
+    "typescript": ">=4.7.0"
```
Comment (iclanton) [inline eslint/eslint-config/package.json:27]: Is this actually necessary? Can we avoid making this a breaking change?

File: heft-plugins/heft-dev-cert-plugin/src/index.ts [hunk 30]
```diff
@@ -12,6 +12,6 @@ import type { IHeftPlugin } from '@rushstack/heft';
 import { DevCertPlugin } from './DevCertPlugin';
 
 /**
- * @internal
+ * @public
```
Comment (iclanton) [inline heft-plugins/heft-dev-cert-plugin/src/index.ts:15]: Why was this made public instead of `@internal`?

Comment (iclanton) [reply heft-plugins/heft-dev-cert-plugin/src/index.ts:15]: Same with the other Heft plugins.

**Review by iclanton — CHANGES_REQUESTED** [review]

File: build-tests/install-test-workspace/workspace/common/pnpm-lock.yaml (hunk 6 above)
Comment (iclanton) [reply build-tests/install-test-workspace/workspace/common/pnpm-lock.yaml:1678]: Yeah this is fine.

File: eslint/eslint-config/package.json (hunk 29 above)
Comment (BPapp-MS) [reply eslint/eslint-config/package.json:27]: This seems like a proper time to do it, and I don't know any way around it- if the consumer of this is using an older version of typescript then Eslint breaks

File: heft-plugins/heft-dev-cert-plugin/src/index.ts (hunk 30 above)
Comment (BPapp-MS) [reply heft-plugins/heft-dev-cert-plugin/src/index.ts:15]: @iclanton It used to be the case that it was marked as internal here and public in the api.md files, due to some strange undocumented behavior. Initially, this change fixed it so that the api.md files were listing these as internal and you said they should be public, so I made the change here to make them public. We have to decide whether we want to change it here to public or in the api.md files to internal

File: common/changes/@rushstack/eslint-config/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json [hunk 31]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Upgrading TypeScript dependency to 4.7, requires new peer dependency of >=4.7",
```
Comment (iclanton) [inline common/changes/@rushstack/eslint-config/benpapp-upgrade-ts-4.7_2022-08-02-20-58.json:4]:
```suggestion
      "comment": "Upgrade TypeScript dependency to 4.7. This package now requires a peerDependency on TypeScript >=4.7",
```

**Review by iclanton — APPROVED** [review]
